apps/cli/src/lib/ProjectTools.ts

```diff
@@ -0,0 +1,54 @@
+import dotenv from "dotenv";
+import * as fs from "node:fs/promises";
+import { DisployConfig, readConfig } from "./disployConf";
+let config: DisployConfig | undefined;
+
+async function resolveProject({ cwd }: { cwd: string }) {
+  if (!config) {
+    const pkg = await readConfig(cwd);
+    config = pkg;
+  }
+
+  return config;
+}
+
+async function findClosestEnv() {
+  let cwd = process.cwd();
+
+  while (cwd !== "/") {
+    const envPath = `${cwd}/.env`;
+    try {
+      await fs.access(envPath);
+      return envPath;
+    } catch (e) {
+      cwd = cwd.substring(0, cwd.lastIndexOf("/"));
+    }
+  }
+
+  return undefined;
+}
+
+async function resolveEnvironment({ cwd }: { cwd: string }) {
+  dotenv.config({ path: await findClosestEnv() });
+
+  const {
+    DISCORD_TOKEN: token,
+    DISCORD_PUBLIC_KEY: publicKey,
+    DISCORD_CLIENT_ID: clientId,
+  } = process.env;
+
+  if (!token || !publicKey || !clientId) {
+    throw new Error("Missing environment variables");
+  }
```

Comment:
Maybe some more specific diagnostics here (like which variables are missing) might be helpful here.

Reply:
resolved in [7ef9f1c](https://github.com/Disploy/disploy/pull/19/commits/7ef9f1cd1ffae01afc8805060f6ecede5694e522)